### PR TITLE
Add `ModuleCacher` for HPU graphs

### PR DESCRIPTION
**Pull request by sungchul1**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algorithms/classification/adapters/mmcls/task.py [hunk 1]
```diff
@@ -69,6 +70,9 @@
 )
 from .utils import build_classifier
 
+if is_hpu_available:
```
Comment (eunwoosh) [inline src/otx/algorithms/classification/adapters/mmcls/task.py]:
```suggestion
if is_hpu_available():
```
It seems typo

File: src/otx/algorithms/segmentation/adapters/mmseg/task.py [hunk 2]
```diff
@@ -63,6 +64,9 @@
 from otx.api.serialization.label_mapper import label_schema_to_bytes
 from otx.api.usecases.tasks.interfaces.export_interface import ExportType
 
+if is_hpu_available:
```
Comment (eunwoosh) [inline src/otx/algorithms/segmentation/adapters/mmseg/task.py]:
```suggestion
if is_hpu_available():
```
Same as above

File: src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_vision_transformer_head.py [hunk 3]
```diff
@@ -38,3 +38,7 @@ def post_process(self, pred):
         """Post processing."""
         pred = cast_bf16_to_fp32(pred)
         return super().post_process(pred)
+
+    def forward(self, x):
```
Comment (eunwoosh) [inline src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_vision_transformer_head.py:42]: It's just a question. Does error occur if this function doesn't exist?

Comment (sungchul1) [reply src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_vision_transformer_head.py:42]: Yes, forward is required.
```
Traceback (most recent call last):
  File "/home/sungchul/workspace/src/training_extensions/venv/cls/bin/otx", line 33, in <module>
    sys.exit(load_entry_point('otx', 'console_scripts', 'otx')())
  File "/home/sungchul/workspace/src/training_extensions/src/otx/cli/tools/cli.py", line 77, in main
    results = globals()[f"otx_{name}"]()
  File "/home/sungchul/workspace/src/training_extensions/src/otx/cli/tools/train.py", line 192, in main
    return train(exit_stack)
  File "/home/sungchul/workspace/src/training_extensions/src/otx/cli/tools/train.py", line 290, in train
    task.train(
  File "/home/sungchul/workspace/src/training_extensions/src/otx/algorithms/classification/task.py", line 213, in train
    results = self._train_model(dataset)
  File "/home/sungchul/workspace/src/training_extensions/src/otx/algorithms/classification/adapters/mmcls/task.py", line 377, in _train_model
    htcore.hpu.ModuleCacher(max_graphs=10)(model=model.head, inplace=True)
  File "/home/sungchul/workspace/src/training_extensions/venv/cls/lib/python3.8/site-packages/habana_frameworks/torch/hpu/graphs.py", line 902, in __call__
    self.forward_params = GraphModel.process_function_signature(self.orig_model.forward)
  File "/home/sungchul/workspace/src/training_extensions/venv/cls/lib/python3.8/site-packages/habana_frameworks/torch/hpu/graphs.py", line 766, in process_function_signature
    assert func_parameters[key].kind not in UNSUPPORTED, \
AssertionError: Unsupported argument type : VAR_POSITIONAL
Exception ignored in: <function ModuleCacher.__del__ at 0x7fba3d82b670>
Traceback (most recent call last):
  File "/home/sungchul/workspace/src/training_extensions/venv/cls/lib/python3.8/site-packages/habana_frameworks/torch/hpu/graphs.py", line 945, in __del__
  File "/home/sungchul/workspace/src/training_extensions/venv/cls/lib/python3.8/site-packages/habana_frameworks/torch/core/weight_sharing.py", line 69, in wrapped__getattr__
  File "/home/sungchul/workspace/src/training_extensions/venv/cls/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1617, in __getattr__
AttributeError: 'ModuleCacher' object has no attribute 'verbose'
```

**Review by eunwoosh — APPROVED** [review]
Thanks for your work!

File: src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_vision_transformer_head.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_vision_transformer_head.py:42]: https://mmpretrain.readthedocs.io/en/stable/api/generated/mmpretrain.models.heads.VisionTransformerClsHead.html#mmpretrain.models.heads.VisionTransformerClsHead.forward

Actually, VisionTrancformerHead should have a forward method
I'm not sure, that self.simple_test is the right choice for forward function

Comment (sungchul1) [reply src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_vision_transformer_head.py:42]: You should check mmcls version adapted on OTX. 0.25.0 doesn't have any forward function, even for parent classes.
https://github.com/open-mmlab/mmpretrain/blob/v0.25.0/mmcls/models/heads/vision_transformer_head.py
https://github.com/open-mmlab/mmpretrain/blob/v0.25.0/mmcls/models/heads/cls_head.py
https://github.com/open-mmlab/mmpretrain/blob/v0.25.0/mmcls/models/heads/base_head.py

And other otx custom heads have that format forward function, such as
https://github.com/openvinotoolkit/training_extensions/blob/d240eba0671520b88f6db0b1f13bb5ed0e68a0b4/src/otx/algorithms/classification/adapters/mmcls/models/heads/custom_cls_head.py#L39-L41

**Pull request merged**